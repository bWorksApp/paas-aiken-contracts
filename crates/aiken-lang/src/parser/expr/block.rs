use chumsky::prelude::*;

use crate::{
    expr::UntypedExpr,
    parser::{error::ParseError, token::Token},
};

pub fn parser(
    sequence: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    choice((
        sequence
            .clone()
            .delimited_by(just(Token::LeftBrace), just(Token::RightBrace)),
        sequence.clone().delimited_by(
            choice((just(Token::LeftParen), just(Token::NewLineLeftParen))),
            just(Token::RightParen),
        ),
    ))
    .map_with_span(|e, span| {
        if matches!(e, UntypedExpr::Assignment { .. }) {
            UntypedExpr::Sequence {
                location: span,
                expressions: vec![e],
            }
        } else {
            e
        }
    })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn block_let() {
        assert_expr!(
            r#"
            let b = {
              let x = 4
              x + 5
            }
            "#
        );
    }

    #[test]
    fn block_single() {
        assert_expr!(
            r#"{
            foo
            }
            "#
        );
    }
}
