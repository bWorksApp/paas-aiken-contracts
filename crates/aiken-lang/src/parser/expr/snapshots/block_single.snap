---
source: crates/aiken-lang/src/parser/expr/block.rs
description: "Code:\n\n{\nfoo\n}\n"
---
Var {
    location: 2..5,
    name: "foo",
}
